### PR TITLE
Blog post on TF TPU training

### PR DESCRIPTION
**Pull request by sayakpaul**
We recently added https://github.com/huggingface/transformers/tree/main/examples/tensorflow/language-modeling-tpu

This blog post is for aiding with the communications around it. 

If you'd instead review the draft on Notion, here's the link:
https://www.notion.so/huggingface2/Training-a-language-model-with-Transformers-using-TensorFlow-and-TPUs-1b61f85e59924f38a5953cda12cbae03

Cc: @Rocketknight1

### CODE REVIEW

File: tf_tpu.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+Unlike our Colab example, however, this example is designed to be **scalable** and much closer to a realistic training run -- although we only use a BERT-sized model by default, the code could be expanded to a much larger model and a much more powerful TPU pod slice by changing a few configuration options.
+
+## Motivation
+
+Why are we writing this guide now? After all, 🤗 Transformers has had support for TensorFlow for several years now. But getting those models to train on TPUs has been a major pain point for the community. This is because:
+
+- Many models weren’t XLA-compatible
+- Data collators didn’t use native TF operations
+
+We think XLA is the future: It’s the core compiler for JAX, it has first-class support in TensorFlow, and you can even use it from PyTorch [link]. As such, we’ve made a [big push](https://blog.tensorflow.org/2022/11/how-hugging-face-improved-text-generation-performance-with-xla.html) to make our codebase XLA compatible, and to remove any other roadblocks standing in the way of XLA and TPU compatibility. This means that users should be able to train most of our TensorFlow models on TPUs without hassle.
```
Comment (gante) [inline tf_tpu.md:31]:
```suggestion
We think XLA is the future: It’s the core compiler for JAX, it has first-class support in TensorFlow, and you can even use it from PyTorch [link]. As such, we’ve made a [big push](https://blog.tensorflow.org/2022/11/how-hugging-face-improved-text-generation-performance-with-xla.html) to make our codebase XLA compatible, and to remove any other roadblocks standing in the way of XLA and TPU compatibility. This means that users should be able to train most of our TensorFlow models on TPUs without a hassle.
```
(missing "a" ?)

**Review by gante — APPROVED** [review]
👌 🔥 🚀 

(Let's make TF great again)

File: tf_tpu.md (hunk 1 above)
Comment (Rocketknight1) [reply tf_tpu.md:31]: Both are valid English, I think! (But I don't actually know why, I just have this vague native speaker intuition)

**Comment by sayakpaul** [comment]
Need to add the publication date and "Edit GitHub" button. Don't merge until that.

**Comment by sayakpaul** [comment]
I have made the necessary changes and have also previewed the blog post. Looking good. Once @osanseviero approves, we should be good to merge.

**Review by Rocketknight1 — APPROVED** [review]
LGTM! (A lot of this has already been through rounds of edits in Notion)

File: tf_tpu.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+## What to expect
+
+We’re going to train a [RoBERTa](https://huggingface.co/docs/transformers/model_doc/roberta) (base model) from scratch on the [WikiText dataset (v1)](https://huggingface.co/datasets/wikitext). As well as training the model, we’re also going to train the tokenizer, tokenize the data and upload it to Google Cloud Storage in TFRecord format, where it’ll be accessible for TPU training. You can find all the code in [this directory](https://github.com/huggingface/transformers/tree/main/examples/tensorflow/language-modeling-tpu). If you’re a certain kind of person, you can skip the rest of this blogpost and just jump straight to the code. If you stick around, though, we’ll take a deeper look at some of the key ideas in the codebase.
```
Comment (osanseviero) [inline tf_tpu.md:37]:
```suggestion
We’re going to train a [RoBERTa](https://huggingface.co/docs/transformers/model_doc/roberta) (base model) from scratch on the [WikiText dataset (v1)](https://huggingface.co/datasets/wikitext). As well as training the model, we’re also going to train the tokenizer, tokenize the data and upload it to Google Cloud Storage in TFRecord format, where it’ll be accessible for TPU training. You can find all the code in [this directory](https://github.com/huggingface/transformers/tree/main/examples/tensorflow/language-modeling-tpu). If you’re a certain kind of person, you can skip the rest of this blog post and just jump straight to the code. If you stick around, though, we’ll take a deeper look at some of the key ideas in the codebase. 
```

File: tf_tpu.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+Since the dataset is already available on the Hub in a compatible format, we can easily load and interact with it using 🤗 datasets. However, for this example, since we’re also training a tokenizer from scratch, here’s what we did:
+
+- Loaded the `train` split of the WikiText using 🤗 datasets.
+- Leveraged 🤗 tokenizers to train a [Unigram model](https://huggingface.co/course/chapter6/7?fw=pt).
+- Uploaded the trained tokenizer on the Hub.
+
+You can find the tokenizer training code [here](https://github.com/huggingface/transformers/tree/main/examples/tensorflow/language-modeling-tpu#training-a-tokenizer) and the tokenizer [here](https://huggingface.co/tf-tpu/unigram-tokenizer-wikitext). This script also allows you to run it with [any compatible dataset](https://huggingface.co/datasets?task_ids=task_ids:language-modeling) from the Hub. 
+
+> 💡 It’s easy to use 🤗 datasets to host your own text datasets. Refer to [this guide](https://huggingface.co/docs/datasets/create_dataset) to learn more.
```
Comment (osanseviero) [inline tf_tpu.md:61]:
```suggestion
> 💡 It’s easy to use 🤗 datasets to host your text datasets. Refer to [this guide](https://huggingface.co/docs/datasets/create_dataset) to learn more.
```

File: tf_tpu.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+## Tokenizing the data and creating TFRecords
+
+Once the tokenizer is trained, we can use it on all the dataset splits (`train`, `validation`, and `test` in this case) and create TFRecord shards out of them. Having the data splits spread across multiple TFRecord shards helps with massively parallel processing as opposed to having each split in single TFRecord files. 
+
+We tokenize the samples individually. We then take a batch of samples, concatenate them together, and split them into several chunks of a fixed size (128 in our case). We follow this strategy rather than tokenizing a batch of samples with a fixed length to avoid aggressively discarding text content (because of truncation). 
+
+We then take these tokenized samples in batches and serialize those batches as multiple TFRecord shards, where the number of shards is determined by the total dataset length and individual shard size. Finally, these shards are pushed to a [Google Cloud Storage (GCS) bucket](https://cloud.google.com/storage/docs/json_api/v1/buckets).
```
Comment (osanseviero) [inline tf_tpu.md:69]:
```suggestion
We then take these tokenized samples in batches and serialize those batches as multiple TFRecord shards, where the total dataset length and individual shard size determine the number of shards. Finally, these shards are pushed to a [Google Cloud Storage (GCS) bucket](https://cloud.google.com/storage/docs/json_api/v1/buckets).
```

File: tf_tpu.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+`DataCollatorForLanguageModeling` is responsible for masking randomly selected tokens from the input sequence and preparing the labels. By default, we return the results from these collators as NumPy arrays. However, many collators also support returning these values as TensorFlow tensors if we specify `return_tensor="tf"`. This was crucial for our data pipeline to be compatible with TPU training.  
+
+Thankfully, TensorFlow provides seamless support for reading files from a GCS bucket:
+
+```python
+training_records = tf.io.gfile.glob(os.path.join(args.train_dataset, "*.tfrecord"))
+```
+
+Here if `args.dataset` contains the `gs://` identifier, TensorFlow will understand that it needs to look into a GCS bucket. Loading locally is as easy as removing the `gs://` identifier.  For the rest of the data pipeline related code, you can refer to [this section](https://github.com/huggingface/transformers/blob/474bf508dfe0d46fc38585a1bb793e5ba74fddfd/examples/tensorflow/language-modeling-tpu/run_mlm.py#L186-#L201) in the training script.
```
Comment (osanseviero) [inline tf_tpu.md:114]:
```suggestion
If `args.dataset` contains the `gs://` identifier, TensorFlow will understand that it needs to look into a GCS bucket. Loading locally is as easy as removing the `gs://` identifier. For the rest of the data pipeline-related code, you can refer to [this section](https://github.com/huggingface/transformers/blob/474bf508dfe0d46fc38585a1bb793e5ba74fddfd/examples/tensorflow/language-modeling-tpu/run_mlm.py#L186-#L201) in the training script.  
```

**Review by osanseviero — APPROVED** [review]
Nice, informative, and fun blog post! I left some minor comments but looks good to be merged

File: tf_tpu.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+- Data collators didn’t use native TF operations
+
+We think XLA is the future: It’s the core compiler for JAX, it has first-class support in TensorFlow, and you can even use it from PyTorch [link]. As such, we’ve made a [big push](https://blog.tensorflow.org/2022/11/how-hugging-face-improved-text-generation-performance-with-xla.html) to make our codebase XLA compatible, and to remove any other roadblocks standing in the way of XLA and TPU compatibility. This means that users should be able to train most of our TensorFlow models on TPUs without hassle.
+
+There’s also another important reason to care about TPU training right now, though: Due to the ongoing orgy of hype and venture capital that followed in the wake of GPT-4’s release, it’s become incredibly hard for most people to get access to state-of-the-art GPUs. Knowing how to train on TPU gives you another path to access ultra-high-performance compute hardware, which is much more dignified than losing a bidding war for the last H100 on eBay and then ugly crying at your desk. You deserve better. And speaking from experience: Once you get comfortable with training on TPU, you might not want to go back.
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tf_tpu.md:33]: @Rocketknight1

**Pull request merged**